Add EIP-6170: Cross-Chain Messaging Interface

sujithsomraaj opened the pull request:
When opening a pull request to submit a new EIP, please use the suggested template: https://github.com/ethereum/EIPs/blob/master/eip-template.md

We have a GitHub bot that automatically merges some PRs. It will merge yours immediately if certain criteria are met:

 - The PR edits only existing draft PRs.
 - The build passes.
 - Your GitHub username or email address is listed in the 'author' header of all affected PRs, inside <triangular brackets>.
 - If matching on email address, the email address is the one publicly listed on your GitHub profile.

abcoathup commented:
Assume this won't be EIP6172 as the first PR was #6170

sujithsomraaj commented:
Should this EIP be EIP6170 then ?

abcoathup commented:
> Should this EIP be EIP6170 then ?

I assume so, but you need an EIP editor to assign a number.

sujithsomraaj commented:
i'm not sure how it works though, when will an EIP editor look at this ?

Pandapip1 commented on eip-6170.md:
```diff
@@ -0,0 +1,79 @@
+
+## Motivation
+
+Cross-chain messaging protocols lack standardization leading to diverse implementation detail. Some examples include Layerzero, Hyperlane & Wormhole, having their own sets of implementations. This makes integration difficult at the aggregator or plugin layer for protocols that must conform to any standards and forces each protocol to implement its adapter.
```
```suggestion
Cross-chain messaging protocols lack standardization, resulting in unnecessarily complex competing implementations: Layerzero, Hyperlane & Wormhole each use a different interface. This makes integration difficult at the aggregator or plugin layer for protocols that must conform to any standards and forces each protocol to implement its adapter.
```

Pandapip1 commented on eip-6170.md:
```diff
@@ -0,0 +1,79 @@
+
+Cross-chain messaging protocols lack standardization leading to diverse implementation detail. Some examples include Layerzero, Hyperlane & Wormhole, having their own sets of implementations. This makes integration difficult at the aggregator or plugin layer for protocols that must conform to any standards and forces each protocol to implement its adapter.
+
+Even chain native arbitrary messaging protocols like Matic State Tunnel has its own way of implementation.
```
```suggestion
Even chain-native arbitrary messaging protocols like the MATIC State Tunnel has an application-specific interface.
```

Pandapip1 commented on eip-6170.md:
```diff
@@ -0,0 +1,79 @@
+
+### Methods
+
+sendMessage
+Dispatch message/state to `receiver` on a predetermined `chain id`.
+MUST emit `MessageSent` event.
+MUST support global chain-ids. Propreitory chain ids should be mapped internally to global chain ids.
+MUST revert if `chain id` is invalid.
+MUST return `true` on successful response from underlying messaging protocol.
+
+**NOTE:** I'm proposing chain-ids be the bytes encoding of their native token name string. For eg., abi.encode("ETH"), abi.encode("SOL") imagining they cannot override.
+       
+       function sendMessage(bytes chainId, bytes receiver, bytes message) public returns(bool)
+
+receiveMessage
+Receive message/state from `sender` on a predetermined `chain id`
+MUST emit `MessageReceived` event.
+MUST support global based chain-ids.
+MUST revert if `sender` or `chain-id` is invalid.
+MUST return `true` on successful processing.
+
+**NOTE:** Sender validation (or) message validation, gas overrides are handled at relayer level.
+
+       function receiveMessage(bytes chainId, bytes sender, bytes message) public returns(bool)
+
+### Events
+
+MessageSent
+MUST trigger when message is sent, include zero bytes transfers.
+       
+       event MessageSent(bytes _sender, bytes _fromChainId, bytes _receiver, bytes _toChainId, bytes message)
+
+MessageReceived
+MUST trigger on any successful call to `receiveMessage(bytes chainId, bytes sender, bytes message)` function. 
+       
+       event MessageReceived(bytes _sender, bytes _fromChainId, bytes _receiver, bytes _toChainId, bytes message)
```
The formatting here is missing. Would you mind e.g. putting this in a solidity interface with NatSpec comments?

sujithsomraaj replied:
sure will do

Pandapip1 commented on eip-6170.md:
```diff
@@ -0,0 +1,79 @@
+
+## Security Considerations
+
+Fully permissionless messaging could be a security threat to the protocol. It is recommended that all the integrators review the implementation of messaging tunnels before integrating.
```
> Fully permissio'nless messaging could be a security threat to the protocol.

How so?

sujithsomraaj replied:
For eg., without sender authentication, anyone could write arbitrary messages into the receiving smart contract. 

This EIP focuses only on the way the messages should be sent and received with a specific standard. But any authentication (or) message tunnel specific operations should be made inside the receive function and should be handled by the integrators or through wrapper contracts.

Pandapip1 replied:
Please put that in the EIP.

Pandapip1 commented on eip-6170.md:
```diff
@@ -0,0 +1,79 @@
+
+## Copyright
+
+Copyright and related rights waived via [CC0](https://eips.ethereum.org/LICENSE)
```
```suggestion
Copyright and related rights waived via [CC0](../LICENSE.md)
```
@SamWilsn how did EIPW not catch this either?

Pandapip1 commented on eip-6170.md:
```diff
@@ -0,0 +1,79 @@
+Cross-chain messaging protocols lack standardization leading to diverse implementation detail. Some examples include Layerzero, Hyperlane & Wormhole, having their own sets of implementations. This makes integration difficult at the aggregator or plugin layer for protocols that must conform to any standards and forces each protocol to implement its adapter.
+
+Even chain native arbitrary messaging protocols like Matic State Tunnel has its own way of implementation.
+
+The lack of standardization not only makes integration difficult but also is error prone as developers need to understand the operation of the messaging protocol at its core.
```
```suggestion
```
The benefits of standardization are well-known to people that read standards.

Pandapip1 commented on eip-6170.md:
```diff
@@ -0,0 +1,79 @@
+Even chain native arbitrary messaging protocols like Matic State Tunnel has its own way of implementation.
+
+The lack of standardization not only makes integration difficult but also is error prone as developers need to understand the operation of the messaging protocol at its core.
+
+A standard for cross-chain messaging will lower integration efforts for cross-chain messaging while creating more robust & secure cross-chain application development.
```
```suggestion
```

Pandapip1 commented on eip-6170.md:
```diff
@@ -0,0 +1,79 @@
+The lack of standardization not only makes integration difficult but also is error prone as developers need to understand the operation of the messaging protocol at its core.
+
+A standard for cross-chain messaging will lower integration efforts for cross-chain messaging while creating more robust & secure cross-chain application development.
+
+To the worst, each messaging protocol has its chain ids (or) chain identification mechanism. However, this mechanism is used as they support non-EVM chains, which makes integration a nightmare.
```
```suggestion
```

Pandapip1 commented on eip-6170.md:
```diff
@@ -0,0 +1,79 @@
+---
+eip: 6170
+title: Cross Chain Messaging Interface
```
```suggestion
title: Cross-Chain Messaging Interface
```

mattstam commented on eip-6170.md:
```diff
@@ -0,0 +1,107 @@
+    /// @param _toChainId is the identifier of the receiving blockchain.
+    /// @param _message is the arbitray message being sent.
+    /// NOTE: this emits when the message is sent, it does not gurantee successful delivery.
+    event MessageSent(
```
These parameters will often want `index` for filtering, and this tends to be significant for off-chain functionality. So much so that folks will [emit two events](https://github.com/hyperlane-xyz/hyperlane-monorepo/blob/b450181c7fc255149ce86551f21dacd7ab47d5f5/solidity/contracts/Mailbox.sol#L84) just to get around the 3 index cap.

mattstam commented on eip-6170.md:
```diff
@@ -0,0 +1,107 @@
+
+## Rationale
+
+The Cross-Chain interface is designed to be optimized for messaging layer integrators with a feature complete yet minimal interface. Validations such as sender authentication, receiver whitelisting, relayer mechanisms and cross-chain execution overrides are intentionally not specified, as Messaging protocols are expected to be treated as black boxes on-chain and inspected off-chain before use.
```
"interoperability layer" from "messaging layer" (the more frequently used term to describe these)
"feature-complete, yet" from "feature complete yet"

```suggestion
The Cross-Chain interface is designed to be optimized for interoperability layer integrators with a feature-complete, yet minimal interface. Validations such as sender authentication, receiver whitelisting, relayer mechanisms and cross-chain execution overrides are intentionally not specified, as Messaging protocols are expected to be treated as black boxes on-chain and inspected off-chain before use.
```

mattstam commented on eip-6170.md:
```diff
@@ -0,0 +1,107 @@
+    /// NOTE: this emits when the message is sent, it does not gurantee successful delivery.
+    event MessageSent(
+        bytes _sender,
+        bytes _fromChainId,
```
`fromChainId` may not be needed on either of these events - if you're listening to these being emitted, you know exactly what chain you're connected to.

mattstam reviewed (changes requested):
The events will be harder to keep consistent between interoperability layer integrations. Consider just having a base interface for `sendMessage()` and `receiveMessage()` only, with a separate interface for events.

Pandapip1 reviewed (changes requested):
LGTM

mattstam commented:
This is looking better, but it is still going to be hard to get multiple interoperability layer integrators to use the same standard - especially on the events, as filtering matters and relayer design varies significantly. 

So I still suggest reducing the scope to just the send/receive functions and asking integrators their thoughts on what they'd like to see for those headers.

mattstam reviewed (commented):
This is looking better, but it is still going to be hard to get multiple interoperability layer integrators to use the same standard - especially on the events, as filtering matters and relayer design varies significantly. 

So I still suggest reducing the scope to just the send/receive functions and asking integrators their thoughts on what they'd like to see for those headers.

sujithsomraaj commented:
> This is looking better, but it is still going to be hard to get multiple interoperability layer integrators to use the same standard - especially on the events, as filtering matters and relayer design varies significantly.
> 
> So I still suggest reducing the scope to just the send/receive functions and asking integrators their thoughts on what they'd like to see for those headers.

Updated for now and removed the events to discuss further with integrators

Pandapip1 reviewed (dismissed):
Misclicked. Meant to set approve instead of request changes.

sujithsomraaj commented:
@Pandapip1 any ideas why the Link checker is failing, is that an issue ?

Pandapip1 commented:
There's a typo. I'll fix it.

Pandapip1 commented:
Fixed. You put the EIP file in the wrong folder.
